include/fmt/format.h

```diff
@@ -1822,85 +1843,201 @@ class specs_handler: public specs_setter<typename Context::char_type> {
 };
 
 // An argument reference.
-template <typename Char>
+template <typename Char, typename String>
 struct arg_ref {
   enum Kind { NONE, INDEX, NAME };
+  typedef Char char_type;
 
-  FMT_CONSTEXPR arg_ref() : kind(NONE), index(0) {}
-  FMT_CONSTEXPR explicit arg_ref(unsigned index) : kind(INDEX), index(index) {}
-  explicit arg_ref(basic_string_view<Char> nm) : kind(NAME) {
-    name = {nm.data(), nm.size()};
-  }
+  FMT_CONSTEXPR arg_ref() : kind(NONE), val() {}
+  FMT_CONSTEXPR explicit arg_ref(unsigned index) : kind(INDEX), val(index) {}
+  FMT_CONSTEXPR explicit arg_ref(String name) : kind(NAME), val(name) {}
 
   FMT_CONSTEXPR arg_ref &operator=(unsigned idx) {
     kind = INDEX;
-    index = idx;
+    val.index = idx;
     return *this;
   }
 
   Kind kind;
-  union {
+  FMT_UNRESTRICTED_UNION value {
+    FMT_CONSTEXPR value() : index(0u) {}
+    FMT_CONSTEXPR value(unsigned id) : index(id) {}
+    FMT_CONSTEXPR value(String n) : name(n) {}
+
     unsigned index;
-    string_value<Char> name;  // This is not string_view because of gcc 4.4.
-  };
+    String name;
+  }
+  val;
 };
 
 // Format specifiers with width and precision resolved at formatting rather
 // than parsing time to allow re-using the same parsed specifiers with
 // differents sets of arguments (precompilation of format strings).
 template <typename Char>
 struct dynamic_format_specs : basic_format_specs<Char> {
-  arg_ref<Char> width_ref;
-  arg_ref<Char> precision_ref;
+  typedef arg_ref<Char, string_value<Char>> arg_ref_type;
+  arg_ref_type width_ref;
+  arg_ref_type precision_ref;
 };
 
 // Format spec handler that saves references to arguments representing dynamic
 // width and precision to be resolved at formatting time.
-template <typename ParseContext>
-class dynamic_specs_handler :
-    public specs_setter<typename ParseContext::char_type> {
+template <typename Specs, typename ParseContext, typename ArgRefCreator>
+class dynamic_specs_handler
+    : public specs_setter<typename ParseContext::char_type> {
```

Comment:
I suggest privately inheriting from `ArgRefCreator` to make use of empty base calls optimization (this will also require making `dynamic_arg_ref_creator` stateless - see comment below).

Reply:
> Because of that dynamic_specs_handler is not able to inherit from its ArgRefCreator.

I don't think this is the case. Base class gives you a subobject in the same way as a member does, it's just optimized away if empty, so you should be able to do:

```c++
...
class dynamic_specs_handler
    : private NameRefCreator, public specs_setter<typename ParseContext::char_type> {
 public:
  FMT_CONSTEXPR dynamic_specs_handler(
      Specs &specs, ParseContext &ctx,
      NameRefCreator name_ref_creator = NameRefCreator())
      : NameRefCreator(name_ref_creator),
        specs_setter<char_type>(specs),
        specs_(specs),
        context_(ctx) {}
  ...
};
```

Reply:
Ahh, I didn't think that it can be implemented like that. Good idea, will do!